lib/ai/model/tokencount.go

```diff
@@ -0,0 +1,273 @@
+// will be used. In this case, the streaming routine must add streamed
+// completion result with the Add() method and call Finish() once the
+// completion is finished. TokenCount() will hang until either Finish() is
+// called or the context is Done.
+type AsynchornousTokenCounter struct {
+	completion *strings.Builder
+	count      int
+
+	// mutex protects all fields of the AsynchornousTokenCounter, it must be
+	// acquired before any read or write operation.
+	// When the count is not finished yet, mutex must be released before waiting.
+	mutex sync.Mutex
+	// finished tells if the count is finished or not.
+	// if the count is already finished, Add() and Finish() will fail.
+	finished bool
+	// waiting is a list of channels of callers wanting to be notified by the
+	// end of the token count. The channels MUST be buffered and have at least 1
+	// available capacity to avoid blocking Finish()
+	waiting []chan struct{}
+}
+
+// TokenCount implements the TokenCounter interface.
+// If the count is already finished, it immediately returns with the count.
+// If the count is not yet finished, it registers a channel to be notified when
+// Finish() is called and waits until either Finish() is called or the context
+// is done.
+func (tc *AsynchornousTokenCounter) TokenCount(ctx context.Context) (int, error) {
+	// If the count is already finished, we return the values
+	tc.mutex.Lock()
+	if tc.finished {
+		defer tc.mutex.Unlock()
+		return tc.count, nil
+	}
+
+	// Else we register and wait for the count to be finished
+	// The channel is buffered with a capacity of one, to avoid the sender to
+	// block if the listener gave up because of a canceled context.
+	waitC := make(chan struct{}, 1)
+	tc.waiting = append(tc.waiting, waitC)
+	tc.mutex.Unlock()
+
+	select {
+	case <-waitC:
+		// No need to acquire lock here because we know the count already
+		// happened and the values cannot change anymore because tc.finished is
+		// set.
+		return tc.count, nil
+	case <-ctx.Done():
+		return 0, trace.Wrap(ctx.Err())
+	}
+}
+
+// Finish marks the token counting process as over. Once finished is called, the
+// token count is computed and the count cannot change anymore. Finish also
+// notifies all waiters that the count is now available.
+func (tc *AsynchornousTokenCounter) Finish() error {
+	tc.mutex.Lock()
+	defer tc.mutex.Unlock()
+	if tc.finished {
+		return trace.Errorf("Already finished counting tokens")
+	}
+
+	// We compute the final count
+	completionTokens, _, err := defaultTokenizer.Encode(tc.completion.String())
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	tc.count = perRequest + len(completionTokens)
+	tc.finished = true
+
+	// And notify everyone that the job is done
+	for _, waiter := range tc.waiting {
+		waiter <- struct{}{}
+	}
+	return nil
+}
+
+// Add a streamed completion response bit to the to-be-counted result.
+func (tc *AsynchornousTokenCounter) Add(delta string) error {
```

Comment:
AFAIK OpenAI API always returns the response token by token. If that's the case, you can just count how many times `Add()` was called instead of constructing the string and counting the tokens after.